### PR TITLE
Devices for metrics and health

### PR DESCRIPTION
**Pull request by lawik**
The start of Device health reporting!

This adds Load avg, memory use, and CPU temp to the Device show page. The data we receive, and store, from the device is a lot more complete, but we have kept it simple for now.

A few bullet points:
- the device health data is requested from Hub once an hour (configurable)
- when you are viewing a device it will start requesting health data once a minute until you leave the device page
- for the time being, only 48 device reports are kept for a device

https://github.com/nerves-hub/nerves_hub_web/assets/8701/c1217367-41ba-4d6d-a5a1-485229bd43f7

### CODE REVIEW

**Comment by lawik** [comment]
fyi @joshk 

I want to test it with `poser` and a real device, it still needs some tests.

I think exposing it in UI can happen separately unless I get frisky :D

File: lib/nerves_hub_web/templates/device/index.html.heex [hunk 1]
```diff
@@ -117,6 +117,21 @@
       <button class="btn btn-secondary" type="button" phx-click="reset-filters">Reset Filters</button>
     </div>
   <% end %>
+
+  <div :if={@health} class="device-health tight">
```
Comment (joshk) [inline lib/nerves_hub_web/templates/device/index.html.heex:121]: can you add a screenshot and show what this looks like on the index page?

Comment (lawik) [reply lib/nerves_hub_web/templates/device/index.html.heex:121]: Screenshot 2024-07-08 at 15 51 25

**Comment by lawik** [comment]
Devices detail view:

Screenshot 2024-07-08 at 15 53 25

File: lib/nerves_hub_web/templates/device/index.html.heex (hunk 1 above)
Comment (joshk) [reply lib/nerves_hub_web/templates/device/index.html.heex:121]: I would skip this view for the time being. It doesn’t feel right here but would work well on a new product dashboard

**Comment by joshk** [comment]
> Devices detail view:
> 
> Screenshot 2024-07-08 at 15 53 25

🔥 🔥 🔥 

this is so awesome!!!

File: lib/nerves_hub_web/templates/device/index.html.heex (hunk 1 above)
Comment (lawik) [reply lib/nerves_hub_web/templates/device/index.html.heex]: Latest update drops this part.

**Comment by joshk** [comment]
I think before we merge this in there needs to be a way to disable health reporting. I can add that tomorrow.

File: lib/nerves_hub/workers/device_health_truncation.ex [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+  The number of days to keep is configured using the environment
+  variable `HEALTH_CHECK_DAYS_TO_RETAIN`
+  """
+
+  use Oban.Worker,
+    max_attempts: 1,
+    queue: :truncate
+
+  @impl true
```
Comment (jjcarstens) [inline lib/nerves_hub/workers/device_health_truncation.ex:13]:
```suggestion
  @impl Oban.Worker
```

File: lib/nerves_hub/devices/device_health.ex [hunk 3]
```diff
@@ -0,0 +1,24 @@
+defmodule NervesHub.Devices.DeviceHealth do
+  use Ecto.Schema
+
+  import Ecto.Changeset
+
+  alias NervesHub.Devices.Device
+
+  alias __MODULE__
```
Comment (jjcarstens) [inline lib/nerves_hub/devices/device_health.ex:8]:
```suggestion
```

File: lib/nerves_hub/devices/device_health.ex [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  @required_params [:device_id, :data]
+
+  schema "device_health" do
+    belongs_to(:device, Device)
+    field(:data, :map)
+    timestamps(type: :utc_datetime_usec, updated_at: false)
+  end
+
+  def save(params) do
+    %DeviceHealth{}
```
Comment (jjcarstens) [inline lib/nerves_hub/devices/device_health.ex:20]:
```suggestion
    %__MODULE__{}
```

**Review by jjcarstens — APPROVED** [review]
Just some no-comment style preferences while reviewing with Josh on a call. Take them of leave them 👍🏼

File: lib/nerves_hub/devices/device_health.ex [hunk 5]
```diff
@@ -0,0 +1,23 @@
+defmodule NervesHub.Devices.DeviceHealth do
+  use Ecto.Schema
+
+  import Ecto.Changeset
+
+  alias NervesHub.Devices.Device
+
```
Comment (joshk) [inline lib/nerves_hub/devices/device_health.ex:7]:
```suggestion
```

Comment (joshk) [reply lib/nerves_hub/devices/device_health.ex:7]: Fix a `mix format` warning

**Pull request merged**